lib/validators.rb

```diff
@@ -137,9 +137,107 @@ def check_shipments_specificities
+      relation_vehicles.each_with_index{ |vehicle_trip, v_i|
+        case v_i
+        when 0
+          unless vehicle_trip[:end_point_id]
+            raise OptimizerWrapper::DiscordantProblemError.new('First trip should at least have end point id')
+          end
+        when relation_vehicles.size - 1
+          unless vehicle_trip[:start_point_id]
+            raise OptimizerWrapper::DiscordantProblemError.new('Last trip should at least have start point id')
+          end
+        else
+          unless vehicle_trip[:start_point_id] && vehicle_trip[:end_point_id]
+            raise OptimizerWrapper::DiscordantProblemError.new('Trip should have start and end point id')
+          end
+        end
+
+        next if v_i.zero? ||
+                vehicle_trip[:start_point_id] == relation_vehicles[v_i - 1][:end_point_id] ||
+                (@hash[:points].find{ |pt| pt[:id] == vehicle_trip[:start_point_id] }[:location] ==
+                  @hash[:points].find{ |pt| pt[:id] == vehicle_trip[:end_point_id] }[:location])
+
+        raise OptimizerWrapper::DiscordantProblemError.new('One trip should start where previous trip ended')
+      }
+    end
+
+    def check_vehicle_trips_timewindows_consistency(relation_vehicles)
+      vehicles_timewindows =
+        relation_vehicles.collect{ |v| v[:timewindow] ? [v[:timewindow]] : v[:sequence_timewindows].to_a }
+
+      return if vehicles_timewindows.all?(&:empty?)
+
+      week_days, unavailable_indices = vehicle_days_availability(relation_vehicles, vehicles_timewindows)
+      if week_days.uniq.size > 1 ||
+         unavailable_indices.uniq.size > 1
+        raise OptimizerWrapper::DiscordantProblemError.new(
+          'Vehicles in vehicle_trips relation have different available days'
+        )
+      end
+
+      vehicles_timewindows.each_with_index{ |v_tw, v_i|
+        next if v_i.zero?
+
+        v_tw.each{ |tw|
+          day = tw[:day_index]
+          previous_tw =
+            vehicles_timewindows[v_i - 1].select{ |ptw|
+              ptw[:day_index].nil? ||
+                ptw[:day_index] == day
+            }.min_by{ |ptw| ptw[:start] }
+
+          unless tw[:end] > previous_tw[:start]
+            raise OptimizerWrapper::DiscordantProblemError.new('Timewindows do not allow vehicle trips')
+          end
+        }
+      }
+    end
+
     def check_relations(periodic_heuristic)
       return unless @hash[:relations].to_a.any?
 
+      @hash[:relations].group_by{ |relation| relation[:type] }.each{ |type, relations|
+        case type
+        when 'vehicle_trips'
+          relations.each{ |relation|
+            relation_vehicles =
+              relation[:linked_vehicle_ids].to_a.collect{ |v_id| @hash[:vehicles].find{ |v| v[:id] == v_id } }
+
+            if relation_vehicles.empty?
+              raise OptimizerWrapper::DiscordantProblemError.new(
+                'Non empty list of vehicles should be provided for every vehicle_trips relations'
```

Comment:
**A** non empty list of vehicle~~s~~ **ids** should be provided for ~~every~~ vehicle_trips relations